relation_prediction/preprocessing.py

```diff
@@ -0,0 +1,116 @@
+import argparse
+import os
+import random
+import networkx as nx
+import numpy as np
+
+
+def _gen_relation2id(G, outdir):
+    outpath = os.path.join(outdir, 'relation2id.txt')
+    relations = set()
+    edges = G.edges()
+    print('{} edges in G'.format(len(edges)))
+    for e in edges:
+        for k, v in G.get_edge_data(*e).items():
+            relations.add(v['relation_id'])
+    relation2id = dict()
+    rid = 0
+    for r in relations:
+        relation2id[r] = rid
+        rid += 1
+    print('{} relations in G'.format(len(relation2id)))
+    print('Writing relation2id to {}'.format(outpath))
+    f = open(outpath, 'a')
+    for k, v in relation2id.items():
+        f.write("{}\t{}\n".format(k, v))
+    print('Done writing relation2id to {}'.format(outpath))
+    f.close()
+
+
+def _gen_entity2id(nodes, outdir):
+    outpath = os.path.join(outdir, 'entity2id.txt')
+    print('{} nodes in G'.format(len(nodes)))
+    entity2id = dict()
+    eid = 0
+    for e in nodes:
+        entity2id[e] = eid
+        eid += 1
+    assert len(nodes) == len(entity2id)
+    print('Writing entity2id to {}'.format(outpath))
+    f = open(outpath, 'a')
+    for k, v in entity2id.items():
+        f.write("{}\t{}\n".format(k, v))
+    print('Done writing entity2id to {}'.format(outpath))
+    f.close()
+
+
+def _write_to_file(path, data):
+    f = open(path, 'a')
+    for x in data:
+        f.write(x)
+    f.close()
+    print('Done writing data to {}'.format(path))
+
+
+def _gen_datasplits(G, outdir, trainprop=.8):
+    trainpath = os.path.join(outdir, 'train.txt')
+    validpath = os.path.join(outdir, 'valid.txt')
+    testpath = os.path.join(outdir, 'test.txt')
+
+    examples = list()
+    for e in G.edges():
+        for k, v in G.get_edge_data(*e).items():
+            ex = '{}\t{}\t{}\n'.format(e[0], v['relation_id'], e[1])
+            examples.append(ex)
+
+    random.shuffle(examples)
+    valprop = (1.-trainprop)/2.
```

Comment:
I am more used to seeing the validation set to have the same size of the test set instead of training set. Not a big deal tho

Reply:
`valprop` is the same as `testprop` here. For example, if trainprop is 0.8, then valprop = (1. - .8)/2 = 0.1. The remaining 0.1 is used for testprop.

Reply:
Ah my mistake